### PR TITLE
CFE-2924 Add inventory of AWS EC2 linux instances

### PR DESCRIPTION
**Pull request by nickanderson**
merge with https://github.com/cfengine/nova/pull/1279

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thank you for submitting a PR! Maybe @craigcomstock can review this?

**Comment by atsaloli** [comment]
@nickanderson I'm surprised to see you used an external shell out to curl to collect instance metadata, why not use libcurl?  Isn't that more lightweight?

For example:

```
bundle agent main
{
  vars:
    any::
      "options_str" string => '
{
  "url.max_content": 1024,
}';
      "options" data => parsejson($(options_str));
      "url" string => "http://169.254.169.254/latest/dynamic/instance-identity/document";
      "result" data => url_get($(url), options);
    ec2_instance::
      "content" data => parsejson("$(result[content])");

  classes:
      "ec2_instance"
        expression => strcmp("$(result[success])", "true");

  reports:
    ec2_instance::
      "imageId = $(content[imageId])";
      "instanceId = $(content[instanceId])";
      "region = $(content[region])";
      "instanceType = $(content[instanceType])";
    !ec2_instance::
      "I am not an AWS EC2 instance. Sob!";
}
```

**Comment by nickanderson** [comment]
> @nickanderson I'm surprised to see you used an external shell out to curl to collect instance metadata, why not use libcurl? Isn't that more lightweight?

I wrote this policy long ago before url_get() was available.  I didn't think about switching it over. One reason not to do so in the currently maintained branches is that we are still maintaining compatibility with 3.7.x for upgrade. After this gets backported perhaps an update to master for 3.13.0 is a good idea.

**Comment by nickanderson** [comment]
```
      "content" data => parsejson("$(result[content])");
```

**Comment by atsaloli** [comment]
Okay :)

**Comment by nickanderson** [comment]
@cf-bottom jenkins, please.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1447/)

https://ci.cfengine.com/job/pr-pipeline/1447/

File: inventory/any.cf [hunk 1]
```diff
@@ -435,6 +443,118 @@ bundle agent cfe_autorun_inventory_cpuinfo
...
+        expression => fileexists( $(cfe_autorun_inventory_aws_ec2_metadata_cache.cache) );
+
+  vars:
+
+    have_cached_instance_identity.ec2_instance.!(disable_inventory_aws|disable_inventory_aws_ec2_metadata)::
+
+      "data" data => readjson( $(cfe_autorun_inventory_aws_ec2_metadata_cache.cache), 100K);
+
+      "region" string => "$(data[region])", meta => { "inventory", "attribute_name=EC2 region" };
+      "instanceId" string => "$(data[instanceId])", meta => { "inventory", "attribute_name=EC2 instanceId" };
```
Comment (olehermanse) [inline inventory/any.cf]: I think the user facing attribute name should be `EC2 Instance ID`(no camelcase). This matches how it's represented in the AWS EC2 Management console. Same for others below.

Comment (nickanderson) [reply inventory/any.cf]: OK, thanks. Will do.

And yes, ack the commit history cleanup.

Comment (nickanderson) [reply inventory/any.cf]: DONE

**Comment by nickanderson** [comment]
Attribute names changed as suggested. Commits squashed.

**Comment by tzz** [comment]
> I wrote this policy long ago before url_get() was available. I didn't think about switching it over. One reason not to do so in the currently maintained branches is that we are still maintaining compatibility with 3.7.x for upgrade. After this gets backported perhaps an update to master for 3.13.0 is a good idea.

You could make it conditional based on the feature class?

File: inventory/any.cf [hunk 2]
```diff
@@ -435,6 +443,118 @@ bundle agent cfe_autorun_inventory_cpuinfo
...
+#
+# Provides:
+#  ec2_instance class based on Amazon markers in dmidecode's system-uuid, bios-version or bios-vendor
+{
+  classes:
+    !disable_inventory_aws::
+      "ec2_instance" -> { "CFE-2924" }
+        comment => "See http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html",
+        scope => "namespace",
+        expression => regcmp("^[eE][cC]2.*", "$(cfe_autorun_inventory_dmidecode.dmi[system-uuid])"),
```
Comment (tzz) [inline inventory/any.cf]: Would this be cleaner as a comparison of `EC2` to `string_upcase(string_head($(cfe_autorun_inventory_dmidecode.dmi[system-uuid]), 3)`? I'm just wondering...

**Comment by tzz** [comment]
Good stuff!!!

File: inventory/any.cf (hunk 2 above)
Comment (nickanderson) [reply inventory/any.cf]: @atsaloli or @mikeweilgart may have a better view on it. If so, take care of it with a follow-up PR please.

**Comment by nickanderson** [comment]
> You could make it conditional based on the feature class?

Thank you for reminding me about the if feature macro!

https://docs.cfengine.com/docs/3.12/reference-macros.html#features

**Comment by nickanderson** [comment]
> Good stuff!!!

Thanks for taking the time to review it!

**Comment by atsaloli** [comment]
👍

File: inventory/any.cf (hunk 2 above)
Comment (atsaloli) [reply inventory/any.cf]: @tzz looks like LISP code ;)

Comment (tzz) [reply inventory/any.cf]: YES IT DOES :P

**Pull request merged**